### DISCUSSION TITLE
High CPU and Memory Load on single node k3s cluster

### DISCUSSION THREAD

**Discussion by philipp1992** [k3s-io/k3s#5769] [General]
HI,

i have an ubuntu 20.02 machine, which has k3s installe. After installing rancher, the memory and cpu consumption of the k3s service is skyrocketing.

versions:
"Ubuntu 20.04.1 LTS"

`- export INSTALL_K3S_VERSION=v1.19.2+k3s1; curl -sfL https://get.k3s.io | sh -`

```
PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
715 root      20   0 4148060   3.0g  52692 S 100.0  38.9  30:03.70 k3s-server
```

the server running the k3s is a normal x86 vserver with 4 cores and 8 gb ram.

```yaml
 # install k3s
    - export INSTALL_K3S_VERSION=v1.19.2+k3s1; curl -sfL https://get.k3s.io | sh -

    # install helm 
    - curl https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3 | bash
    - starting=true ;while $starting ; do kubectl get ns && starting=false;sleep 1;done

    # install cert-manager
    - kubectl create namespace cert-manager
    - helm repo add jetstack https://charts.jetstack.io
    - helm repo update
    - KUBECONFIG=/etc/rancher/k3s/k3s.yaml helm install cert-manager jetstack/cert-manager --wait --namespace cert-manager --version v1.0.2 --set installCRDs=true

    # install rancher
    - helm repo add rancher-latest https://releases.rancher.com/server-charts/latest
    - helm repo update
    - kubectl create namespace cattle-system
    - KUBECONFIG=/etc/rancher/k3s/k3s.yaml helm install rancher rancher-latest/rancher  --wait --namespace cattle-system --set replicas=1
```
```
● k3s.service - Lightweight Kubernetes
     Loaded: loaded (/etc/systemd/system/k3s.service; enabled; vendor preset: enabled)
     Active: active (running) since Mon 2020-10-12 09:58:20 CEST; 10min ago
       Docs: https://k3s.io
    Process: 2277 ExecStartPre=/sbin/modprobe br_netfilter (code=exited, status=0/SUCCESS)
    Process: 2279 ExecStartPre=/sbin/modprobe overlay (code=exited, status=0/SUCCESS)
   Main PID: 2280 (k3s-server)
      Tasks: 191
     Memory: 3.3G
```

any idea what could cause the high load?

kind regards
Philipp

**Comment by brandond** [comment]
I'm not able to reproduce this on a local KVM server with the same specs - Ubuntu 20.04 x86_64, 4 cores / 8GB RAM.

I do see that the CPU and memory utilization increase and use ~2 cores while Rancher is installing, but it settles down once all the pods are running. The only difference is that I haven't used LetsEncrypt since my dev node is not exposed to the internet.

I will note that my node itself is on a 10.0.0.0/8 network so it is allowed to talk to itself via those allow rules, despite the fact that the normal k3s ports have not been exposed via UFW. What address space are you using for your node?

install:
```bash
echo "y" | ufw enable
ufw allow ssh
ufw allow http
ufw allow https
ufw default deny incoming
ufw default allow outgoing
ufw allow from any to 10.0.0.0/8
ufw allow from 10.0.0.0/8 to any

export INSTALL_K3S_VERSION=v1.19.2+k3s1; curl -sfL https://get.k3s.io | sh -

curl https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3 | bash
helm repo add jetstack https://charts.jetstack.io
helm repo add rancher-latest https://releases.rancher.com/server-charts/latest
helm repo update

kubectl create namespace cert-manager
KUBECONFIG=/etc/rancher/k3s/k3s.yaml helm install cert-manager jetstack/cert-manager --wait --namespace cert-manager --version v1.0.2 --set installCRDs=true

kubectl create namespace cattle-system
KUBECONFIG=/etc/rancher/k3s/k3s.yaml helm install rancher rancher-latest/rancher  --wait --namespace cattle-system --set replicas=1 --set hostname=ubuntu01.lan.khaus
```

kubectl get nodes -o wide:
```
root@ubuntu01:~# kubectl get nodes -o wide
NAME       STATUS   ROLES    AGE   VERSION        INTERNAL-IP   EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION     CONTAINER-RUNTIME
ubuntu01   Ready    master   18m   v1.19.2+k3s1   10.0.1.147    <none>        Ubuntu 20.04.1 LTS   5.4.0-48-generic   containerd://1.4.0-k3s1
```

kubectl get pods -A:
```
NAMESPACE                 NAME                                       READY   STATUS      RESTARTS   AGE
kube-system               metrics-server-7b4f8b595-pgb4z             1/1     Running     0          13m
kube-system               local-path-provisioner-7ff9579c6-cbj2h     1/1     Running     0          13m
kube-system               coredns-66c464876b-bq6z6                   1/1     Running     0          13m
kube-system               helm-install-traefik-cvw8c                 0/1     Completed   0          13m
kube-system               svclb-traefik-tnmt6                        2/2     Running     0          13m
kube-system               traefik-5dd496474-zdp4t                    1/1     Running     0          13m
cert-manager              cert-manager-cainjector-5f988f74c6-jmllv   1/1     Running     0          11m
cert-manager              cert-manager-57b65b7fc-v5w5f               1/1     Running     0          11m
cert-manager              cert-manager-webhook-7cf554f879-tdkzk      1/1     Running     0          11m
cattle-system             rancher-7797675cb7-4szmj                   1/1     Running     0          10m
fleet-system              gitjob-68cbf8459-hcn8w                     1/1     Running     0          8m43s
fleet-system              fleet-controller-5db6bcbb9-prf9l           1/1     Running     0          8m43s
cattle-system             helm-operation-9dllc                       0/2     Completed   0          8m54s
cattle-system             helm-operation-n6x4z                       0/2     Completed   0          8m32s
rancher-operator-system   rancher-operator-6659bbb889-hqfrk          1/1     Running     0          8m25s
cattle-system             helm-operation-2sdrn                       0/2     Completed   0          8m27s
fleet-system              fleet-agent-7584f654d9-sg2kc               1/1     Running     0          8m10s
cattle-system             rancher-webhook-6f69f5fd94-qjdvx           1/1     Running     0          8m10s
cattle-system             helm-operation-2kxhq                       0/2     Completed   0          8m12s
cattle-system             helm-operation-h5tqz                       0/2     Completed   0          8m3s
```

top:
```
    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
   1183 root      20   0 2412884   1.6g  92228 S  13.0  20.2   3:29.67 /usr/local/bin/k3s server
   4174 root      20   0 1025056 303564  71080 S   1.0   3.8   0:13.73 rancher --http-listen-port=80 --https-listen-port=443 --audit-log-path=/var/log/auditlog/ran+
   1198 root      20   0  893332 148572  71228 S   0.3   1.8   0:36.39 containerd
   3401 sysadm    20   0  806308  94092  22756 S   0.3   1.2   0:03.06 /app/cmd/cainjector/cainjector --v=2 --leader-election-namespace=kube-system
   5023 root      20   0  759568  78972  38532 S   0.0   1.0   0:02.59 fleetcontroller
   6715 root      20   0  758972  74308  36792 S   0.3   0.9   0:01.18 fleetagent
```

**Reply by Alex2357** [reply]
Do you wanna say 13% on idle server without any load is normal? See my post above according to times it used 36 CPU hours in 8h.

**Comment by brandond** [comment]
Can you confirm the version you're running? The process name (k3s-server) in your top output is not what I would expect to see. What do you get from `kubectl get nodes -o wide` and `kubectl get pods -A`

**Comment by risayew** [comment]
Hi @brandond , Hi @philipp1992 . I have the same issue.
top:

```
top - 21:58:36 up 48 days,  5:17,  1 user,  load average: 6.31, 6.68, 6.91
Tasks: 168 total,   1 running, 118 sleeping,   0 stopped,   0 zombie
%Cpu(s):  6.2 us,  1.2 sy,  0.0 ni, 68.9 id, 23.6 wa,  0.0 hi,  0.2 si,  0.0 st
KiB Mem :  4038784 total,   116712 free,  2991612 used,   930460 buff/cache
KiB Swap:  4038652 total,  3455996 free,   582656 used.  1275144 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                                                           
  560 root      20   0 1258400 1.061g  81772 S  24.6 27.5   7:58.63 k3s-server                                                                                                                        
17432 mysql     20   0 5089688 259096   6660 S   6.6  6.4   2507:05 mysqld    

```
kubectl get nodes -o wide :

```
NAME       STATUS   ROLES    AGE   VERSION        INTERNAL-IP     EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION       CONTAINER-RUNTIME
mynode   Ready    master   48d   v1.18.9+k3s1   192.168.0.2   <none>        Ubuntu 18.04.4 LTS   4.15.0-112-generic   containerd://1.3.3-k3s2
```

kubectl get pods -A

```
NAMESPACE           NAME                                                       READY   STATUS      RESTARTS   AGE
kube-system         metrics-server-7566d596c8-2pksv                            1/1     Running     0          48d
kube-system         svclb-traefik-j58rf                                        2/2     Running     0          48d
cattle-prometheus   prometheus-operator-monitoring-operator-5bb7ddcd65-lwf9c   1/1     Running     0          48d
cattle-system       cattle-cluster-agent-7cf8c7bb67-7jtns                      1/1     Running     0          25d
cattle-system       cattle-node-agent-t8gnn                                    1/1     Running     0          25d
cattle-system       system-upgrade-controller-585f9d68f9-nd6hf                 1/1     Running     0          51m
cert-manager        cert-manager-7cb75cf6b4-w6bm6                              1/1     Running     1          48d
kube-system         coredns-7944c66d8d-qfrtl                                   1/1     Running     0          48d
cert-manager        cert-manager-webhook-7c75b89bf6-m9bcj                      1/1     Running     0          48d
kube-system         traefik-758cd5fc85-5s68b                                   1/1     Running     0          48d
kube-system         local-path-provisioner-6d59f47c7-dtcbs                     1/1     Running     5          48d
cert-manager        cert-manager-cainjector-759496659c-hnlmk                   1/1     Running     5          48d
cattle-system       rancher-5745c55b85-6wrgk                                   1/1     Running     0          25d
cattle-system       rancher-5745c55b85-wcdmp                                   1/1     Running     1          25d
cattle-system       rancher-5745c55b85-hc2v5                                   1/1     Running     0          25d
kube-system         helm-install-traefik-gtnm2                                 0/1     Completed   0          47m

```

Very high load averages. And on the 4-node RKE cluster too, which is added to the rancher k3s 1 node cluster.

**Comment by brandond** [comment]
@risayew 25% of one core and ~1GB of RAM is on the high end of normal for k3s, but not out of line for what I'd expect while hosting Rancher, which is pretty hard on the apiserver due to all the custom resources. @philipp1992 reported k3s consistently using more than 100% and 3+ GB of memory which is unusual.

**Comment by risayew** [comment]
Thanks for the answer. And what do you think about load averages? They increase with the time. I have seen even >12. I have such values because the kubernetes was just upgraded several minutes ago to 1.18.9

**Comment by brandond** [comment]
Load average doesn't really correspond to much in terms of resource utilization. There's a reason it's not used for scheduling. Just keep an eye on your CPU and memory.

**Comment by unixfox** [comment]
@risayew 
> Thanks for the answer. And what do you think about load averages? They increase with the time. I have seen even >12. I have such values because the kubernetes was just upgraded several minutes ago to 1.18.9

Which version were you using before?

Am I crazy if I say that I didn't have any high CPU and memory usage on k3s 1.17?

**Comment by risayew** [comment]
@unixfox Thats interesting. The previous version of the Kubernetes was 1.18.8.
On another rancher setup i manage 4 -Nodes Cluster too and have the same rancher version 2.4.8.
And on that 4-Nodes (more or less same specs) Cluster with Kubernetes v1.16.15 i observe way lower cpu/load averages.

**4 Nodes Cluster on v1.18.9**
<img width="797" alt="Screenshot 2020-10-06 at 14 01 20" src="https://user-images.githubusercontent.com/2446408/95965623-6b719600-0e0a-11eb-996d-9249b92350e5.png">

_**VS**_
**4 Nodes Cluster on v1.16.15**
<img width="805" alt="Screenshot 2020-10-06 at 14 07 23" src="https://user-images.githubusercontent.com/2446408/95965792-9fe55200-0e0a-11eb-81e4-9f163ab857b1.png">

May be CPU utilization does not have anything to do with k3s or rancher, but with Kubernetes version.

**Comment by volume-ji** [comment]
I encountered this problem too, repeated created/deleted resources(deployment, job, etc), memory increased continuously.

k3s version: container image  rancher/k3s:v1.16.14-k3s1

**Comment by risayew** [comment]
Additionally i observe too often ETCD downs and leader changes. Somebody has an idea, where should i start to look?

**Comment by brandond** [comment]
@risayewin my experience, excessive storage latency is the usual cause of etcd leader changes.

**Comment by rancher-max** [comment]
I have noticed this before on longer-running clusters, but for me was indicative of something else wrong going on in my cluster, most often this being a pod that was stuck in a perpetual CrashLoopBackoff state. I've tried over the last few days to reproduce this by running a continuous script that would add a few, deployments, pods, and services and then delete them every few seconds, but still did not notice any increase in CPU or memory. 

I was using 2cpus, 4gb ram, 10gb disk and saw my node consistently utilizing approx. 13% cpu and 41% mem with minor fluctuations. These numbers were grabbed from the scenario using v1.19.2+k3s1, with rancher deployed on the cluster.

**Comment by henkesn** [comment]
Looks like this bug has been fixed with https://github.com/rancher/kine/commit/3faf3a7028014a5baf96454b2b3fe04984ebc69f in kine 0.5.1, hasn't it?

**Comment by brandond** [comment]
While Kine does now compact more efficiently, that was not a significant contributor towards CPU or memory utilization. It was mostly just slow with external datastores due to the need to frequently call out to the SQL server during the compaction process.

**Comment by brandond** [comment]
We've added some resource profiling figures to our documentation feel free to take a look:
https://rancher.com/docs/k3s/latest/en/installation/installation-requirements/resource-profiling/

**Comment by transacid** [comment]
Hey, This "bug" seems to be the most relatable one to my issue. Since upgrading to >=1.21 .0 the cpu usage on my raspberry is ~20% higher. I have tried up to v1.21.2+k3s1 but eventually rolled back to v1.20.8+k3s1. I'm on armv7 and I remember that there was a regression with kubernetes and the go version build for armv7 (I can't remember exactly when that was). I had that issue before and was happy when it finally dissipated with a newer version. with 1.21 the issue seems to be back. I'll attach a screenshot of the cpu usage. (The timeframes are quite large because I do not spend too much time up/downgrading).
Upgrade to 1.21:
![upgrade to 1.21](https://user-images.githubusercontent.com/113231/125855263-f45b1845-14d1-40c3-9f3c-fc927117d06c.png)
Downgrade back to 1.20
![downgrade to 1.20](https://user-images.githubusercontent.com/113231/125855313-e2053086-5b9c-4f14-b9cc-8acbb5e91dfb.png)

**Comment by nilbacardit26** [comment]
@transacid thank you for pointing that out. I am having the same issue with v1.21.2+k3s1 deployed as a master. I have a cluster of 4 rapsberry pi 4, and the truth is that I dunno if I am experiencing this after I upgraded my master node to this version or not. The cpu load is extremely high, above 130% using the top command. I will give it a try and then posting the results

**Comment by MrSaints** [comment]
I can confirm that I've had the exact same experience as @transacid. Downgrading immediately rectified the problem. To rule other factors out, I've even attempted a clean install, a reinstall, and a full node restart. All in isolation. The downgrade made the biggest impact.

EDIT: I was wrong. It appears that downgrading only alleviates the problem, but over time, the CPU peaks again.

**Comment by brandond** [comment]
The perf regression in 1.21 is an upstream issue, ref:
* https://github.com/kubernetes/kubernetes/issues/101989

**Comment by a5r0n** [comment]
i have same problem, But with version **v1.19.4+k3s1**
k3s single node (VM on proxmox), with 12 cores and 16GB RAM.

```
Tasks: 628 total,   1 running, 618 sleeping,   0 stopped,   9 zombie
%Cpu(s): 58.9 us, 12.5 sy,  0.0 ni, 22.9 id,  0.1 wa,  0.0 hi,  0.6 si,  5.0 st
MiB Mem :  16022.1 total,   5875.5 free,   8096.4 used,   2050.2 buff/cache
MiB Swap:      0.0 total,      0.0 free,      0.0 used.   7930.5 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                                                                                                                           
  530 root      20   0 2571432   1.4g  49964 S 827.6   9.2  58021:36 k3s-server                                                                                                                                        
  713 root      20   0 1090024 240376  43608 S  23.9   1.5  15813:46 containerd 
```

![Screenshot_20210804_172516](https://user-images.githubusercontent.com/32464596/128198889-a82f9c3d-c7af-4f3f-8576-6a62aa8be27f.png)

**Comment by technicool** [comment]
I'm seeing the same CPU usage creep on a 3 node AWS `t3a.small` cluster (all master's w/ embedded etcd). The cluster was deployed with k3sup for the default 1.21 version of k3s.

Restarting a node would fix the CPU usage, but after about 6-12 hours, it's starting to go from a baseline of 10% up to about 15% CPU.

This cluster is running a few containers that are mostly idle, and running `top` seems to show that the `k3s-server` process is consuming the vast majority.
Also of note, most of the CPU time is actually in `si` or software interrupt, according to `top`'s documentation.

**Comment by brandond** [comment]
@technicool you should note that t3a.small instances (like all burstable instance types) are designed to run at less than 20% average CPU utilization; if you remain above that for an extended period of time you will start getting throttled. I suspect that after 6-12 hours you're running out of CPU credits.

https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/burstable-credits-baseline-concepts.html#:~:text=an%20instance%20stops)-,Earn%20CPU%20credits,-Each%20burstable%20performance

You should be able to monitor this in cloudwatch metrics.

**Comment by technicool** [comment]
@brandond that is correct about 20%, however, the cpu for k3s-server starts around 9%, and increases above 20% over several days. So its not just a credit issue.

 The crazy part is that cpu credits build up for a day or three, then as the CPU starts increasing over time, it starts draining them, eventually starving at the 20% due to AWS throttling

Its always annoying when reboots solve the issue. (Resetting back to the 9% usage level)

**Comment by a5r0n** [comment]
> i have same problem, But with version **v1.19.4+k3s1**
> k3s single node (VM on proxmox), with 12 cores and 16GB RAM.

Yesterday the problem got worse, k8s api became unavailable, and an attempt to restart the k3s service repeatedly failed. (A long time until the database loads, then nothing.)
I found that the database file reached a size of 25G total, and after running a `vacuum` (sqlite),
db size is 4.8G and  k3s was able to run successfully.

The CPU usage is still very high, about 1000%.. reboot didn't help.  WTF 🤕

**Comment by pletnes** [comment]
I've just done a

1. Fresh install of ubuntu server 20.04
2. Fresh install of k3s
3. No k8s/k3s workloads running. No other workloads, either.

I am also seeing 20% CPU usage on an intel i5. If k3s is suitable for raspberry pi level hardware, why does it need 20% constant cpu to do nothing?

Hats off for a super slick installation procedure, though!

**Comment by pcgeek86** [comment]
I am experiencing the same problem on a Raspberry Pi 3B+. k3s uses so much CPU that the device is barely functional. k3s is using about 30-50% of CPU on a relatively fresh install of Ubuntu Server 20.04.3 LTS 64-bit. I can't even run any k3s commands. It just hangs.

**Comment by askoriy** [comment]
I have similar issue. To find root cause I tried deploy with disabled all components:
```
k3s server \
    --disable coredns \
    --disable servicelb \
    --disable traefik \
    --disable local-storage \
    --disable metrics-server \
    --disable-scheduler \
    --disable-cloud-controller \
    --disable-kube-proxy \
    --disable-network-policy \
    --disable-helm-controller 
```
and still got CPU load about 10%.
Disabling of metrics server has the most effect on decreasing CPU usage. (but only decrease LA to 30%)

**Comment by whyman** [comment]
Seeing this on a fresh 22.04 
k3s version v1.23.6+k3s1 (418c3fa8)
go version go1.17.5

**Comment by xavier83** [comment]
can't find a similar issue on k0s though, wonder what they are doing differently :thinking:

**Comment by moio** [comment]
>  Disabling of metrics server has the most effect on decreasing CPU usage. (but only decrease LA to 30%)

That's what I noticed as well. Is it expected?

**Comment by thiscantbeserious** [comment]
Chances are good this is caused by etcd - back in the days I have been observing and testing its performance and reliability - not only is it a resource hog, not designed to run on low power machines at all it has also some bugs deeply embeded in regards to ARM.

Not sure what the state is in regards to that but I would highly recommending not using a multi-master node setup with it but rather something like this (e.g. can be combined with a bare metal Postgres cluster)

https://github.com/alexellis/k3sup#create-a-multi-master-ha-setup-with-external-sql

Bottomline of the story: A Multi-Master Setup will always have performance overhead even in idle since its required to be always in sync across the Masters in case one goes down.

Also if you want maximal performance then dont expect it out of the box. You will need to know what you can and do deploy. You will need to understand the technology involved.

Especially on a complex clustering environment.

**Comment by aliuq** [comment]
I have the same issue, I reinstall k3s in OS more times, it always take high memory over 2g, I trust it works if I reinstall the OS, but I want to find a safe type to clear k3s resource instead of reinstall OS. the last I found it seems `kubectl top nodes` had a few errors

```bash
kubectl top nodes
NAME     CPU(cores)   CPU%   MEMORY(bytes)   MEMORY%
master   73m          3%     2110Mi          57%
```

```bash
free -m
              total        used        free      shared  buff/cache   available
Mem:           3677         973         513           2        2189        2411
Swap:             0           0           0
```

**Comment by Alex2357** [comment]
Same with me Ubuntu 22.04

```
top - 07:15:20 up  9:25,  1 user,  load average: 1.16, 0.65, 0.41
Tasks: 153 total,   1 running, 152 sleeping,   0 stopped,   0 zombie
%Cpu(s):  4.1 us,  8.1 sy,  0.0 ni, 87.8 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem :   7840.3 total,   3782.0 free,    834.1 used,   3224.3 buff/cache
MiB Swap:   2048.0 total,   2048.0 free,      0.0 used.   6738.0 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
 399223 root      20   0 1270076 506052 107816 S  11.8   6.3  38:46.91 k3s-server
 406737 65532     20   0  839280  96468  71872 S  11.8   1.2   0:20.12 traefik
 400105 root      20   0  772820  62000  38328 S   5.9   0.8   5:08.39 containerd
 717232 root      20   0       0      0      0 I   5.9   0.0   0:00.06 kworker/u8:2-events_power_efficient
 720679 root      20   0   13384   4356   3648 R   5.9   0.1   0:00.03 top
```

[v1.25.8-rc1+k3s1](https://github.com/k3s-io/k3s/releases/tag/v1.25.8-rc1%2Bk3s1)
Sha256sum of my airgup install

```
8421e18c844df501a6bc525eaf56e1cbd93278d9b03c1b78db8d576652241cae  k3s
61192992bb798161944ed29f8e7596f8bd87ff5a1b6f8dbbaafeb79c32f21369  k3s-airgap-images-amd64.tar
88152dfac36254d75dd814d52960fd61574e35bc47d8c61f377496a7580414f3  install.sh
```

```
root@k3:~# kubectl top node
NAME   CPU(cores)   CPU%   MEMORY(bytes)   MEMORY%   
mp     452m         11%    2198Mi          28%       
root@k3:~# kubectl top pod -A
NAMESPACE     NAME                                      CPU(cores)   MEMORY(bytes)   
kube-system   coredns-597584b69b-mbsbp                  4m           13Mi            
kube-system   local-path-provisioner-79f67d76f8-25xfv   1m           6Mi             
kube-system   metrics-server-5f9f776df5-mwn59           17m          16Mi            
kube-system   svclb-traefik-8e433bcd-snsj6              0m           0Mi             
kube-system   traefik-66c46d954f-frz87                  1m           25Mi            
root@k3:~# 
```

Was installing by running

```
ssh root@${rhost} "INSTALL_K3S_SKIP_DOWNLOAD=true  sh -s " < install.sh
```

**Comment by ArnCo** [comment]
This is still an issue. I experience the same symptoms as described in this article on 2 out of 3 of my K3S / rancher installs : https://canthonyscott.com/k3s-cpu-usage-bug-remains/ . Indeed, the CPU slowly increases over time.

All of them run:
- Kube v1.25.9 +k3s1
- Rancher v2.7.3
No issue on Ubuntu 22.04.2 LTS on ext4.

Issue is present on my two archlinux installs over btrfs, and etcd is known for messing with btrfs. Anybody experiencing the issue running on this filesystem?
I'll update this thread if it can be of any use...

**Comment by ghost62** [comment]
Hi, how to get this information? 

```
● k3s.service - Lightweight Kubernetes
     Loaded: loaded (/etc/systemd/system/k3s.service; enabled; vendor preset: enabled)
     Active: active (running) since Mon 2020-10-12 09:58:20 CEST; 10min ago
       Docs: https://k3s.io
    Process: 2277 ExecStartPre=/sbin/modprobe br_netfilter (code=exited, status=0/SUCCESS)
    Process: 2279 ExecStartPre=/sbin/modprobe overlay (code=exited, status=0/SUCCESS)
   Main PID: 2280 (k3s-server)
      Tasks: 191
     Memory: 3.3G
```

**Reply by dereknola** [reply]
`sudo systemctl status k3s`

**Reply by ghost62** [reply]
Thank you!!

**Comment by Blackclaws** [comment]
Potentially same issue: https://github.com/k3s-io/k3s/discussions/7786

My issue got moved to discussion as there wasn't a direct idea that it was a k3s bug, however reading this discussion I get the feeling something about k3s must be broken.

**Comment by bobobo1618** [comment]
I'm confused. On a single-node cluster, why is etcd even running? The default datastore is sqlite and yet I see both sqlite and etcd in my logs.

**Reply by brandond** [reply]
Kine implements the etcd APIs. It is etcd, as far as the apiserver is concerned. It is not possible that both would be in use simultaneously.

**Comment by zonque** [comment]
I'm experiencing the same issue with version 1.28.5 and 1.29.1. The node barely has any load, but k3s is eating ~25% of the CPU resources constantly. Any hint on how to debug this?

**Reply by zonque** [reply]
30s CPU profile:

![profile001](https://github.com/k3s-io/k3s/assets/106304/10519ac7-bb6e-4ace-b6c6-4c7d39e5d729)

**Reply by brandond** [reply]
This is expected. See https://github.com/k3s-io/k3s/issues/2278#issuecomment-771104128 and the many other issues/threads opened by folks that expect Kubernetes to use 0 CPU at idle, without understanding that "idle" actually involves significant continuous heartbeating, health checks, and metrics collection.

**Reply by tadghh** [reply]
> This is expected. See [#2278 (comment)](https://github.com/k3s-io/k3s/issues/2278#issuecomment-771104128) and the many other issues/threads opened by folks that expect Kubernetes to use 0 CPU at idle, without understanding that "idle" actually involves significant continuous heartbeating, health checks, and metrics collection.

Im lost on this, why dont I have this idle issue when using minikube? When using the same setup with k3s I can hear my severs fans spin up to 3000rpm every 15min unlike that minikube deployment that didnt have this issue.

**Comment by InsOpDe** [comment]
this just sucks - guess i need so switch k8s distribution.
![image](https://github.com/user-attachments/assets/500b3a13-895c-47d4-8acf-3b7f2eeb7cb5)

**Reply by brandond** [reply]
Or you could figure out why it's doing that. It's not normal, and something is probably wrong/misconfigured.

Note that unless you are running like 50 copies of K3s in Docker, you're looking at threads, not processes.

**Reply by tadghh** [reply]
not op but im pretty sure the usage might be related to the metrics server

**Comment by tadghh** [comment]
Update for anyone else, it might be related to garbage collection. Adding the following to your config.yaml may help.

```yaml
## /etc/rancher/k3s/config.yaml
etcd-compaction-interval: "240m"
kube-controller-manager-arg:
  - "--concurrent-gc-syncs=5"
etcd-cron-compaction-time: "0 2 * * *"

```
